# 문자열(String)



---



## 문자열 정의


* 문자 단어 등으로 구성된 문자들의 집합
* 시퀀스 자료형(sequence data type)

In [2]:
s="String"
print(s)
print(s[0]) #0번째 인덱스 출력
print(s[1])
print(s[2])
print(s[3])
print(s[4])
print(s[5])

String
S
t
r
i
n
g


In [3]:
s='문자열'
print(s) #한글에 대해서도 지원함 유니코드로
print(s[0])
print(s[1])
print(s[2])

문자열
문
자
열


## 문자열 생성

* 파이썬에서 문자열은 작은따옴표(') 또는 큰 따옴표(")로 표현


In [1]:
print('Hello')
print("파이썬은 재미있다.")
print('''파이썬은 심플하다.''')
print("""파이썬은 문자열 처리가 뛰어나다.""")

Hello
파이썬은 재미있다.
파이썬은 심플하다.
파이썬은 문자열 처리가 뛰어나다.


### 따옴표가 있는 문자열 생성

* 문자열에 작은따옴표가 있을 경우, 큰따옴표로 둘러싸서 표현
* 문자열에 큰따옴표가 있을 경우, 작은따옴표로 둘러싸서 표현
* 이스케이프 코드 \ (원화표시=백슬래쉬)를 이용하여 작은따옴표(‘)와 큰따옴표(“)를 문자열에 포함

In [3]:
string="Python's built-in string classes"
print(string) #잘나온다
string='Python's built-in string classes'
print(string) #에러난다

SyntaxError: ignored

In [5]:
string='모두가 "파이썬을 매우 쉽다."라고 말한다.'
print(string)
string='Python\'s built-in string classes' #이스케이프 문자를 써서 ' 뒤에도 문자열로 인식되게 한다.
print(string)
string = "모두가 \"파이썬을 매우 쉽다.\"라고 말한다." # 안쪽의 쌍따옴표, 이스케이프 바로 뒤에 있는 문자가 이스케이프 효과를 받음.
print(string)

모두가 "파이썬을 매우 쉽다."라고 말한다.
Python's built-in string classes
모두가 "파이썬을 매우 쉽다."라고 말한다.


# 이스케이프 문자
* 특수 문자
* 문자 앞에 백슬래쉬를 사용
* 문자열에서 특수 문자를 표현하는데 사용<br>
\<br>
\'<br>
\"<br>
\a<br>
̄̄\b<br>
\f<br>                                     
\n<br> 한줄 내려주는거 라인피드
\r<br>
\t<br>
\v<br>
\ooo<br>
\xhh


\### 여러 줄이 있는 문자열 생성

* 이스케이프 문자(\\n)를 이용하여 여러 줄이 있는 문자열 생성 가능

In [2]:
text = "동해물과 백두산이 마르고 닳도록\n하느님이 보우하사 우리나라 만세\n무궁화 삼천리 화려강산\n대한사람, 대한으로 길이 보전하세."
print(text)

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세
무궁화 삼천리 화려강산
대한사람, 대한으로 길이 보전하세.


* 작은 따옴표 3개 또는 큰 따옴표 3개를 이용하여 여러 줄이 있는 문자열 생성

In [3]:
text = """동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세.
무궁화 삼천리 화려강산
대한사람, 대한으로 길이 보전하세."""
print(text)

동해물과 백두산이 마르고 닳도록
하느님이 보우하사 우리나라 만세.
무궁화 삼천리 화려강산
대한사람, 대한으로 길이 보전하세.




---



## 문자열 연산(String Operators)

### 문자열 더하기

+ 연산자를 사용하여 문자열 연결


In [6]:
s1 = "동해물과 "
s2 = "백두산이"
print(s1 + s2)

동해물과 백두산이


### 문자열 곱하기

* \* 연산자를 사용하여 문자열 반복


In [7]:
s = "String "
print(s * 3)

String String String 


### 문자열 길이(length)

* 문자열 길이를 구하는 `len()` 함수

In [8]:
s = "String"
print(len(s)) # 7개
print(len(s * 3)) # 7개 * 3

6
18


### 문자열 인덱싱(indexing)

* 문자열은 리스트처럼 문자 하나하나가 상대적인 주소(offset)를 가짐
* 이 주소를 사용해 할당된 값을 가져오는 인덱싱을 사용

 S  t  r  i  n  g <br>
 0  1  2  3  4  5 <br>
-6 -5 -4 -3 -2 -1<br>

앞에서는 그냥 알던대로 0 1 2 3 ~~<br>
뒤에서 시작하면 0이 아니라 -1 -2 -3 -4 ~

In [9]:
s = "String"
print(s)
print(s[-6])
print(s[-5])
print(s[-4])
print(s[-3])
print(s[-2])
print(s[-1])

String
S
t
r
i
n
g


### 문자열 슬라이싱(slicing)

* 문자열의 주소를 이용하여 문자열을 조각(부분)을 추출

In [17]:
s = "SuanLab - Suan computer laboratory"
print(s)
print(s[0:7]) #0번째 인덱스부터 6번째 인덱스
print(s[:7]) #0은 생략이 가능 바로 위랑 같음
print(s[10:]) #끝값도 안써 됨. 끝까지 다 출력하겠다 그러면 콜론뒤에 안쓰면 됨
print(s[-10:]) # -10(L)부터 끝까지 출력해라
print(s[-19:-11]) # -19(C)부터 -11(r)까지 출력하기~

SuanLab - Suan computer laboratory
SuanLab
SuanLab
Suan computer laboratory
laboratory
computer




---



## 문자열 메소드(String Methods)


문자열 처리를 쉽게하는 메서드 제공 <br>
capitalize 첫문자 대문자 나머지 소문자<br>
casefold 모든 대소문자 구분 제거<br>
count<br>
find<br>
rfind<br>
index<br>
rindex<br>
isalnum 다 숫자면 true 아니면 false 반환<br>
isalpha<br>
isdecimal<br>
isdigit<br>
isnumeric<br>
isidentifier 문자열이 유효한 식별자인 경우 True<br>
isnumeric<br>
isspace<br>
istitle<br>
islower<br>
isupper<br>
<br>
lower 다 소문자로 바꿔서 반환<br>
upper<br>
title  첫글짜만 대문자 나머지는 소문자 문자열로 반환<br>
swapcase 대 소문자 서로바꿔서 반환<br>
<br>
strip([chars]) 문자열 양쪽의 chars 또는 공백을 제거한 문자열을 반환
<br>

### capitalize() / casefold()

* 문자열 "string"에 대해 `capitalize()`를 실행하면, 첫 문자가 대문자인 "String" 문자열로 변환
* 문자열 "String"에 대해 `casefold()`를 실행하면, 모든 문자가 소문자인 "string" 문자열로 변환

In [19]:
s = "string"
print(s)
s = s.capitalize()
print(s)
s=s.casefold()
print(s)

string
String
string


### count()

* 문자열 "string string"에 포함된 문자 's'의 갯수를 반환하도록 `count()`를 실행
* 문자 's'가 총 2개 존재하므로 2를 반환
* 부분 문자열 "str"이 몇개인지 반환하는 것도 가능하고, 부분 문자열 "str"이 총 2개 존재하므로 2를 반환

In [24]:
s="string string"
print(s)
print(s.count('s')) # s의 갯수가 몇갠지 반환
print(s.count("str")) # str의 갯수가 몇갠지 반환

string string
2
2


### find() / rfind()

* 문자열 "string string"에서 문자 's'의 해당 위치를 `find()`를 통해 반환
* 문자열 "string string"에서 부분 문자열 "ing"의 해당 위치를 `find()`를 통해 반환
* 문자열 "string string"에서 문자 's'를 해당 위치를 `rfind()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 부분 문자열 "ring"를 해당 위치를 `rfind()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환

In [29]:
s='string string'
print(s.find('s')) # s에 대해서 위치값(인덱스)를 반환
print(s.find('ing')) # ing는 인덱스 3에 있음
print(s.rfind('s')) # rfind는 오른쪽부터 찾는거 7번 인덱스에 있다.
print(s.rfind('ing')) # 위와 동일 10 인덱스에 있다.

0
3
7
10


### index() / rindex()

* 문자열 "string string"에서 문자 's'의 해당 위치를 `index()`를 통해 반환
* 문자열 "string string"에서 부분 문자열 "ing"의 해당 위치를 `index()`를 통해 반환
* 문자열 "string string"에서 문자 's'를 해당 위치를 `rindex()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 부분 문자열 "ring"를 해당 위치를 `rindex()`를 통해 오른쪽부터 탐색하여 가장 큰 인덱스를 반환
* 문자열 "string string"에서 문자 'z'를 해당 위치를 `rindex()` 함수를 통해 오른쪽부터 탐색하지만 찾을 수 없어서 'ValueError' 발생

In [30]:
s='string string'
print(s.index('s')) # s에 대해서 위치값(인덱스)를 반환
print(s.index('ing')) # ing는 인덱스 3에 있음
print(s.rindex('s')) # rfind는 오른쪽부터 찾는거 7번 인덱스에 있다.
print(s.rindex('ing')) # 위와 동일 10 인덱스에 있다.
#print(s.rindex('z')) #없는걸 찾으면 밸류에러가 뜬다. 

0
3
7
10


### isalnum()

* `isalnum()`은 문자열에 알파벳이나 숫자가 1개 이상 있으면 True 반환
* 문자열 "string"은 알파벳으로 구성 되었으므로 True 반환
* 문자열 "한글"은 알파벳으로 구성 되어있으므로 True 반환
* 문자열 "!@#"은 특수기호로 구성 되어있으므로 False 반환
* 문자열 "123"은 숫자들로 구성 되어있으므로 True 반환

In [31]:
print("string".isalnum()) # 문자열에 바로 점 찍고 메서드 사용가능
print("한글".isalnum()) # 한글도 유니코드라서 알파벳임.
print("!@#".isalnum()) # 알파벳도 아니고 넘버도 아니다.
print("123".isalnum())

True
True
False
True


### isalpha()

* `isalpha()`는 문자열에 알파벳이 1개 이상 있으면 True 반환
* 문자열 "string"은 알파벳으로 구성 되었으므로 True 반환
* 문자열 "한글"은 알파벳으로 구성 되어있으므로 True 반환
* 문자열 "!@#"은 특수기호로 구성 되어있으므로 False 반환
* 문자열 "123"은 숫자들로 구성 되어있으므로 Flase 반환

In [32]:
print("string".isalpha()) # 문자열에 바로 점 찍고 메서드 사용가능
print("한글".isalpha()) # 한글도 유니크도라서 알파벳임.
print("!@#".isalpha()) # 알파벳도 아니고 넘버도 아니다.
print("123".isalpha())

True
True
False
False


### isdecimal()

* `isdecimal()`는 문자열의 모든 문자가 10진수 문자이면 True 반환
* 문자열 "123"은 모두 10진수 문자열이므로 True 반환
* 문자열 "1.23"은 실수형 문자열이므로 False 반환

In [37]:
print("string".isdecimal()) 
print("한글".isdecimal()) 
print("!@#".isdecimal()) 
print("123".isdecimal()) # 얘만 10진수 숫자
print("1.23".isdecimal()) # 실수는 안됨

False
False
False
True
False


### isdigit()

* `isdigit()`는 문자열의 모든 문자가 숫자일 때 True 반환
* 문자열 "123"은 모든 문자가 숫자에 해당하므로 True 반환
* 문자열 "1.23"은 문자 중 숫자에 해당안되는 . 이 존재하여 False

In [39]:
print("string".isdigit()) 
print("한글".isdigit()) 
print("!@#".isdigit()) 
print("123".isdigit()) 
print("1.23".isdigit()) # 점이 껴있어서 false임.

False
False
False
True
False


### isnumeric()

* `isnumeric()`는 문자열의 모든 문자가 수치형일 때 True 반환
* 문자열 "123"은 모든 문자가 숫자에 해당하므로 True 반환
* 문자열 "1.23"은 문자 중 수치형에 해당안되는 . 이 존재하여 False

In [40]:
print("123".isnumeric()) # 다 수치형
print("1.23".isnumeric()) # ,이 수치형이 아니니까 false

True
False


### isidentifier()

* `isidentifier()`는 문자열이 파이썬에서 사용하는 식별자인 경우 True 반환
* 문자열 "123"은 식별자로 사용되지 않는 문자열이라서 False 반환
* 문자열 "True"는 식별자로 사용되기 때문에 True 반환
* 문자열 "print"는 식별자로 사용되기 때문에 True 반환

In [43]:
print("123".isidentifier())
print("True".isidentifier()) # True 이런게 식별자임.
print("Print".isidentifier()) # Print가 식별자임

False
True
True


### isspace()

* 문자열 " "은 문자열 내에 공백 문자가 있어서 True 반환
* 문자열 " 1 "은 문자열 내에 공백 아닌 문자가 있어서 False 반환

In [45]:
print(" ".isspace()) #공백이 있으면 트루
print(" 1".isspace()) #공백이 아닌 문자가 포함돼있어서 펄스

True
False


### istitle()

* 문자열 "String"은 문자열 내에 제목과 같이 첫 글자가 대문자여서 True 반환
* 문자열 "STRING"은 모든 문자가 대문자이기 때문에 False 반환

In [47]:
print("String".istitle()) # 첫자만 대문자면 트루
print("STRING".istitle()) # 다 대문자면 펄스

True
False


### islower()

* 문자열 "string"은 모든 문자가 소문자여서 True 반환
* 문자열 "String"은 첫 문자가 대문자여서 False 반환

In [48]:
print("string".islower())
print("String".islower())

True
False


### isupper() 함수

* 문자열 "STRING"은 모든 문자가 대문자여서 True 반환
* 문자열 "String"은 첫 문자만 대문자여서 False 반환

In [49]:
print("STRING".isupper())
print("String".isupper())

True
False


### join()

* 문자열 "String"에 대해서 `join()`으로 공백 문자 ' '를 문자 사이마다 추가한 문자열 "S t r i n g"을 반환
* 문자열 "String"에 대해서 `join()` 으로 공백 문자 '_'를 문자 사이마다 추가한 문자열 "S_t_r_i_n_g"을 반환
* 문자열 "String"에 대해서 `join()` 으로 공백 문자 '|'를 문자 사이마다 추가한 문자열 "S|t|r|i|n|g"을 반환

In [52]:
s = "String"
print(' '.join(s))  # String 글자 사이사이에 공백이 들어가게 됨
print('_'.join(s))
print('|'.join(s))

S t r i n g
S_t_r_i_n_g
S|t|r|i|n|g


### center() / ljust() / rjust()

* 문자열 "String"에 대해서 `center()`를 사용하여 너비 10에 해당하는 문자열 길이에 가운데 정렬한 문자열 '  String  '로 변환
* 문자열 "String"에 대해서 `ljust()`를 사용하여 너비 10에 해당하는 문자열 길이에 왼쪽 정렬한 문자열 'String    '로 변환
* 문자열 "String"에 대해서 `rjust()`를 사용하여 너비 10에 해당하는 문자열 길이에 오른쪽 정렬한 문자열 '    String'로 변환

In [57]:
print("'" + "String".center(10) + "'") # '과 ' 사이에 String을 가운데 정렬시킴
print("'" + "String".ljust(10) + "'") # 왼쪽으로 정렬
print("'" + "String".rjust(10) + "'") # 오른쪽으로 정렬

'  String  '
'String    '
'    String'


### lower() / upper() / title() / swapcase()

* 문자열 "String"을 `lower()`를 이용하여 모든 문자가 소문자로 변환된 "string" 반환
* 문자열 "String"을 `upper()`를 이용하여 모든 문자가 대문자로 변환된 "STRING" 반환
* 문자열 "string"을 `title()`를 이용하여 첫 문자가 대문자로 변환된 "String" 반환
* 문자열 "String"을 `swapcase()`를 이용하여 대문자는 소문자로 변환되고 소문자는 대문자로 변환된 "sTRING" 반환


In [58]:
print("String".lower())
print("String".upper())
print("String".title())
print("String".swapcase())

string
STRING
String
sTRING


### strip() / lstrip() / rstrip()

* 양쪽 공백이 포함된 문자열 "  String  "을 `strip()`를 통해 공백 제거
* 양쪽 공백이 포함된 문자열 "  String  "을 `lstrip()`를 통해 왼쪽 공백 제거
* 양쪽 공백이 포함된 문자열 "  String  "을 `rstrip()`를 통해 오른쪽 공백 제거

In [60]:
print("   String   ".strip()) # 양쪽공백이 제거된 결과 반환
print("   String   ".lstrip()) # 왼쪽 공백이 제거된 결과 반환
print("   String   ".rstrip()) # 오른쪽 공백이 제거된 결과 반환

String
String   
   String


### partition() / rpartition()

* 문자열 "String"을 `partition()`를 이용하여 문자 't' 기준으로 분할
* 문자열 "String String"을 `rpartition()`를 이용하여 마지막 문자 'S'를 기준으로 분할

In [65]:
print("String".partition('t')) # t를 기준으로 3개로 나눈다.
print("String String".rpartition('S')) # r 이니까 오른쪽에 있는 인덱스 큰 s를 기준으로 파티션

('S', 't', 'ring')
('String ', 'S', 'tring')


### replace()

* 문자열 "String"을 `replace()`를 이용해 문자열 "Str"을 문자 'R'로 교체

In [66]:
print("String".replace("Str", 'R'))

Ring


### split() / rsplit() / splitlines()

* 문자열 "1 2 3"을 `split()`의 기본 구분자를 이용하여 '1', '2', '3' 으로 구분
* 문자열 "1_2_3"을 `split()`으로 구분자 '_'를 이용하여 '1', '2', '3' 으로 구분
* 문자열 "1 2 3"을 `rsplit()`으로 구분자 '_'를 이용하여 오른쪽부터 1번만 구분하여 '1_2', '3' 으로 구분
* 문자열 "123\n123\n123\n"을 `splitlines()` 함수를 이용하여 라인 단위로 구분하여 '123, '123, '123 반환

In [71]:
print("1 2 3".split()) # 123이 1, 2, 3으로 구분됨. 기본구분자(공백)이 사용됨
print("1_2_3".split('_'))
print("1_2_3".rsplit('_', 1)) # r이니까 오른쪽서부터 1, 한번만 스플릿해라.
print("123\n123\n123\n".splitlines()) # 라인단위\n로 스플릿하기

['1', '2', '3']
['1', '2', '3']
['1_2', '3']
['123', '123', '123']


### startswith() / endswith()

* 문자열 "String"에서 시작 문자가 'S'인지 `startswith()`를 통해 확인
* 문자열 "String"에서 마지막 문자가 'g'인지 `endswith()`를 통해 확인

In [74]:
print("String".startswith('S'))
print("String".endswith('g')) #뭘로 시작하고 뭘로 끝나는지 알수 있는 메서드

True
True


### zfill()

* 문자열 "123"에서 너비 8만큼으로 늘리고 비어있는 부분에 '0'이 채워진 문자열 "00000123" 반환

In [76]:
print("123".zfill(8)) # zfill=zerofill 문자열이 123이 있는데 총 8개 글자로 만들고 빈곳은 0으로 채워줘 

00000123




---



## 문자열 서식(String Format)


### 문자열 포맷팅(String Formatting)

* 문자열 내에서 서식에 맞추어 특정 값을 삽입 또는 변경
* 기호 % 뒤에 있는 값이 문자열 내의 서식에 순서대로 매핑

In [83]:
print("나는 사과가 %d개 있다." % 2) # % 뒤라는 위치에 2라는 값이 매핑이 되는것.
print("나는 사과가 %s개 있다." % "두 ")
print("나는 %s가 %d개 있다." % ("배", 5)) # 퍼센트가 두군데면 괄호치고 콤마로 구분

나는 사과가 2개 있다.
나는 사과가 두 개 있다.
나는 배가 5개 있다.


### 정렬, 공백, 소수점 포맷

* 포맷 문자 앞에 숫자는 길이를 의미
* -는 왼쪽 정렬을 의미
* 소수점 ‘.’ 뒤에 숫자는 소숫점 이하 개수를 의미

In [89]:
print("%8s" % "Hello") # 총길이를 8로 해놔서 헬로 앞에 공백 3개 들어감
print("%-8s" % "Hello") # 8개의 길이를 가지는데 -니까 왼쪽정렬
print("%-8sPython" % "Hello") # %-8 Hello를 먼저 인식하고(8자 왼쪽정렬) 써놓고 그담에 파이썬을 씀
print("%0.2f" % 3.1415926535)
print("%8.2f" % 3.1415926535)
print("%-8.2f" % 3.1415926535)

   Hello
Hello   
Hello   Python
3.14
    3.14
3.14    


### 문자열 format 메소드

* 문자열 `format()`의 인덱스를 이용한 고급 포맷팅
* 인덱스 번호에 따라 `format()` 함수의 각 값들이 매핑

In [91]:
print("나는 사과가 {0}개 있다.".format(2))
print("나는 사과가 {0}개 있다.".format("두"))
print("나는 {0}가 {1}개 있다.".format("사과", 2))

나는 사과가 2개 있다.
나는 사과가 두개 있다.
나는 사과가 2개 있다.


* 문자열 `format()` 메소드의 인덱스와 변수 이름을 이용한 포맷팅
* 변수 이름 s와 n을 지정하여 "사과"와 숫자 2를 매핑

In [92]:
print("나는 사과가 {n}개 있다.".format(n=2))
print("나는 {s}가 {n}개 있다.".format(s="사과", n=2))

나는 사과가 2개 있다.
나는 사과가 2개 있다.


* 문자열 `format()` 함수의 정렬, 공백, 소수점


In [95]:
print("{0:8}".format("Hello"))
print("{0:<8}".format("Hello")) #왼쪽정렬
print("{0:>8}".format("Hello")) #오른쪽정렬
print("{0:^8}".format("Hello")) #가운데 정렬
print("{0:~^8}".format("Hello")) #~로 공백들을 채우기

Hello   
Hello   
   Hello
 Hello  
~Hello~~


### f 문자열 포맷팅

* 문자열 포맷팅을 위해 새롭게 나온 기능으로 문자열 앞에 f를 붙여서 구분
* 변수 s와 n에 대해서 문자열에 포함된 이름으로 매핑하여 사용 가능

In [97]:
s = "사과"
n = 2
print(f"나는 사과가 {n}개 있다.") # f로 티를 내는것. 지금 f문자 포맷팅하는거다~
print(f"나는 {s}가 {n}개 있다.") # 앞에 있는 변수를 알아서 심플하게 맵핑해서 사용

나는 사과가 2개 있다.
나는 사과가 2개 있다.


* 기존의 문자열 포맷을 문자열 다음 기호 : 뒤에 표현하여 그대로 사용 가능

In [102]:
print(f'{"Hello":8}')
print(f'{"Hello":<8}')#왼쪽정렬
print(f'{"Hello":>8}') #오른쪽정렬
print(f'{"Hello":^8}') #가운데 정렬
print(f'{"Hello":~^8}') #~로 공백들을 채우기

Hello   
Hello   
   Hello
 Hello  
~Hello~~
